### ISSUE TITLE
SOAP gradients computes superfluous zero entries

### ISSUE THREAD

**Issue by max-veit** [cosmo-epfl/librascal#107]
Exactly what it says on the tin: The current SOAP gradients implementation computes entries that are known to be zero from the species keys (e.g. if taking the gradient of a SOAP entry with species (C, Si) wrt a neighbour of type Ge, the gradient is _known_ to be zero, but it's iterated over and explicitly zeroed anyway). Our first attempt at skipping the zero-gradient entries (a98605efd9) was unsuccessful, so we're going to look into this more closely and fix it soon.

**Comment by Luthaf** [comment]
This code seems to be related (see https://github.com/cosmo-epfl/librascal/pull/150#discussion_r333900165), saving it here for future usage:

```cpp
// Neighbour gradients need a separate pair list because if the species
// of j is not the same as either of the species for that SOAP entry,
// the gradient is zero.
for (auto neigh : center) {
  std::vector<internal::SortedKey<Key_t>> grad_pair_list{};
  for (const auto & el1 : coefficients) {
    auto && neigh_1_type{el1.first[0]};
    for (const auto & el2 : coefficients) {
      auto && neigh_2_type{el2.first[0]};
      auto neigh_type = neigh.get_atom_type();
      if (neigh_1_type <= neigh_2_type) {
        pair_type[0] = neigh_1_type;
        pair_type[1] = neigh_2_type;
        if ((neigh_type == pair_type[0]) or
            (neigh_type == pair_type[1])) {
          grad_pair_list.emplace_back(is_sorted, pair_type);
        }
      }
    }
  }
}
```

**Comment by Luthaf** [comment]
Fixed in #228 as far as I know
